### PR TITLE
MP-0006: Adding proposal for built in properties.

### PR DESCRIPTION
**Pull request by johnament**
Signed-off-by: John D. Ament <john.d.ament@gmail.com>

# Proposal for a MicroProfile specification

**! DO NOT MERGE this Pull Request BEFORE REVIEW is finished!**

**Merging this Pull Request means that the proposal is officially accepted by the MicroProfile Core Team as a MicroProfile incubation project. A public discussion and an agreement within the MicroProfile Core Team is expected before a proposal is accepted/rejected.** 

_If this PR is not a proposal for a new specification, delete this template description._

## Suggested steps to follow before this proposal is accepted ##

 - add label `proposal` to this PR on github
 - initiate a public discussion on the mP mailing list
   - Email Subject: "Review announcement: <<PROPOSAL_NAME>>"
   - Email Body: Based on [review announcement template](https://github.com/microprofile/evolution/blob/master/process.md#review-announcement) 
       - fill in `<<PROPOSAL NAME>>`, `<<REVIEW END DATE>>` (not mandatory, can be announced later, but should be at least 7 days after the announcement email is sent) 
       and `<<REVIEW MANAGER NAME>>` (the name of the sender)
       - update the link to a proposal (NNNN-proposal.md) to point to this Pull Request

After the review, the Core Team should form an agreement whether or not to accept the proposal, or whether ask its author for amendments. 

Once this Pull Reuest is merged, the proposal is considered accepted and an incubation repository should be created to facilitate collaboration.

### CODE REVIEW

File: proposals/0006-standard-properties_pt_1.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+## Introduction
+
+In order to build portable applications and frameworks across different Microprofile runtimes, certain assumptions need to be made about the runtime.  This includes other paired runtimes in addition the availability of certain configuration information.  This spec aims to define some basic properties that may be useful to users as well as framework developers to better integrate with Microprofile.
+
+Mailinglist thread: [Discussion thread topic for that proposal](https://groups.google.com/forum/#!topic/microprofile/e58HZ5cNLGQ)
+
+## Motivation
+
+While we've defined basic server runtime capabilities (CDI + JAX-RS + JSON-B) we have not defined simple things such as that a service is avilable over HTTP.  It is implied with the notion of Microservices.
```
Comment (mikecroft) [inline proposals/0006-standard-properties_pt_1.md]: > CDI + JAX-RS + JSON-B

Should be JSON-P, JSON-B isn't in MP [yet]

**Comment by mikecroft** [comment]
I like the proposal. A standard configuration is great, but a few well-known properties like these would be great.

File: proposals/0006-standard-properties_pt_1.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+## Proposed solution
+
+The proposed solution includes some basic properties that define your server runtime, which may include things like:
+
+- `microprofile.listen.address` - the hostname/IP of where your service listens
+- `microprofile.listen.port` - the port that your service is listening on
+
```
Comment (Emily-Jiang) [inline proposals/0006-standard-properties_pt_1.md:27]: Are the properties required to be provided by the runtime as system properties?

Comment (Emily-Jiang) [reply proposals/0006-standard-properties_pt_1.md:27]: For the properties, we need to mark them as mandatory or optional. I am thinking to introduce more fault tolerance configurations but they are provided  by the applications and optional.

File: proposals/0006-standard-properties_pt_1.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+These properties aren't meant to be exhaustive, but instead represent some basic configuration of your system runtime that others may find useful.  As the spec is actually defined, we would anticipate listing out an exhaustive list for Microprofile 1.1.
+
+Using the configuration API, a developer may be able to read their listen port or address using the following code snippet:
+
+```
+String listenAddress = ConfigProvider.getConfig().getValue("microprofile.listen.address", String.class);
+int listenPort = ConfigProvider.getConfig().getValue("microprofile.listen.port", Integer.class);
+```
+
+Note that these vales would be read only from the user's point of view, there would no need to modify them, meaning placing these properties in a `ConfigSource` would not need to be supported.
```
Comment (NottyCode) [inline proposals/0006-standard-properties_pt_1.md:37]: Is it correct to read this as the developer is not able to set these properties, only read them and the setting of them would still be based on the vendor specific configuration mechanism?

File: proposals/0006-standard-properties_pt_1.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+While we've defined basic server runtime capabilities (CDI + JAX-RS + JSON-B) we have not defined simple things such as that a service is avilable over HTTP.  It is implied with the notion of Microservices.
+
+## Proposed solution
+
+The proposed solution includes some basic properties that define your server runtime, which may include things like:
+
+- `microprofile.listen.address` - the hostname/IP of where your service listens
+- `microprofile.listen.port` - the port that your service is listening on
+
+These properties aren't meant to be exhaustive, but instead represent some basic configuration of your system runtime that others may find useful.  As the spec is actually defined, we would anticipate listing out an exhaustive list for Microprofile 1.1.
```
Comment (NottyCode) [inline proposals/0006-standard-properties_pt_1.md]: In the discuss thread you mention that it isn't targeted for a specific version of micro profile, but here you list a specific version. Wondering how I should correctly interpret these two different statements.

Comment (johnament) [reply proposals/0006-standard-properties_pt_1.md]: Agreed, can change that.

File: proposals/0006-standard-properties_pt_1.md (hunk 3 above)
Comment (johnament) [reply proposals/0006-standard-properties_pt_1.md:37]: I think that sounds right.

File: proposals/0006-standard-properties_pt_1.md (hunk 1 above)
Comment (johnament) [reply proposals/0006-standard-properties_pt_1.md]: Heh.

File: proposals/0006-standard-properties_pt_1.md (hunk 2 above)
Comment (johnament) [reply proposals/0006-standard-properties_pt_1.md:27]: They are intended to be read from Config spec, possibly, or from some other runtime provider.  I have a prototype of what the latter means in [Hammock](https://github.com/hammock-project/hammock/blob/master/core/src/main/java/ws/ament/hammock/HammockRuntime.java), where you can use full type safety to get this data outside of config (though I anticipate replacing DS config with another open source config only solution based on the config spec)

**Comment by kwsutter** [comment]
Looks good and the review has been open for over a week.  I'll merge in the changes.

**Pull request merged**